File: src/core/HotkeyManager.ts
import * as path from 'node:path';
import * as fs from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { ITTSService } from './interfaces/ITTSService.js';
import { AudioService } from '../infrastructure/AudioService.js';
import { ISettingsManager } from './interfaces/ISettingsManager.js';

/**
 * Data shape representing a hotkeyed predefined speech phrase.
 */
export interface HotkeyEntry {
    /** Unique numeric identifier for the hotkey entry. Maps directly to `hotkey_${id}.wav`. */
    id: number;
    
    /** Text sentence associated with this hotkey. Synthesized on demand or pre-rendered. */
    text: string;
    
    /** Accelerator/hotkey key combination configuration string (e.g. 'CommandOrControl+Alt+1') */
    hotkey: string;
}

/**
 * Manages shortcut hotkeys and pre-generated audio clips for instant phrase playback.
 * Minimizes real-time speech generation latencies by pre-rendering audio files.
 */
export class HotkeyManager {
    /** Directory where pre-rendered hotkeyed WAV audios are cached */
    private hotkeyedDir: string;
    
    /** Absolute path to the JSON file indexing assigned hotkeys */
    private jsonPath: string;
    
    /** Internal cache of hotkey entries */
    private entries: HotkeyEntry[] = [];
    
    /** Limit on max hotkeys active simultaneously */
    private readonly MAX_ENTRIES = 20;

    constructor(
        private ttsService: ITTSService,
        private audioService: AudioService,
        private settingsManager: ISettingsManager
    ) {
        this.hotkeyedDir = path.join(settingsManager.getAppDirectory(), 'audio', 'hotkeyed');
        this.jsonPath = path.join(this.hotkeyedDir, 'hotkeys.json');
    }

    /**
     * Initializes hotkey directories and loads metadata from disk.
     */
    public async init(): Promise<void> {
        if (!existsSync(this.hotkeyedDir)) {
            await fs.mkdir(this.hotkeyedDir, { recursive: true });
        }
        if (existsSync(this.jsonPath)) {
            try {
                const data = await fs.readFile(this.jsonPath, 'utf-8');
                this.entries = JSON.parse(data);
            } catch (e) {
                this.entries = [];
            }
        }
    }

    /**
     * Writes hotkeys cache metadata back to disk.
     */
    private async save(): Promise<void> {
        await fs.writeFile(this.jsonPath, JSON.stringify(this.entries, null, 2), 'utf-8');
    }

    /**
     * Associates a keyboard hotkey combination with a specific phrase.
     * Pre-renders and saves the voice phrase immediately to disk to guarantee instant playback when triggered.
     * @param hotkey Key combination accelerator.
     * @param text Sentence to synthesize.
     */
    public async assignHotkey(hotkey: string, text: string): Promise<void> {
        let entryIndex = this.entries.findIndex(e => e.hotkey === hotkey);
        let id: number;

        // If hotkey already exists, reuse its ID and overwrite its text/file
        if (entryIndex >= 0) {
            id = this.entries[entryIndex].id;
            this.entries[entryIndex].text = text;
        } else {
            // Find first available integer ID hole (0 to MAX_ENTRIES - 1)
            const usedIds = new Set(this.entries.map(e => e.id));
            id = -1;
            for (let i = 0; i < this.MAX_ENTRIES; i++) {
                if (!usedIds.has(i)) {
                    id = i;
                    break;
                }
            }

            // Fallback: If no holes found, eject the oldest hotkey entry
            if (id === -1) {
                const oldest = this.entries.shift();
                if (oldest) id = oldest.id;
                else id = 0;
            }

            this.entries.push({ id, text, hotkey });
        }

        const filePath = path.join(this.hotkeyedDir, `hotkey_${id}.wav`);
        console.log(`Assigning hotkey '${hotkey}' to ID [${id}]...`);
        
        // Trigger immediate background synthesis to destination path
        await this.ttsService.generateToFile(text, filePath);
        
        await this.save();
        console.log(`Hotkey '${hotkey}' assigned successfully.`);
    }

    /**
     * Plays a hotkeyed phrase audio file by ID using current audio configuration.
     * @param id Identifier of the hotkey recording.
     */
    public async playHotkey(id: number): Promise<void> {
        const entry = this.entries.find(e => e.id === id);
        if (!entry) {
            console.log(`No audio assigned to hotkey ID: ${id}`);
            return;
        }

        const filePath = path.join(this.hotkeyedDir, `hotkey_${entry.id}.wav`);
        if (!existsSync(filePath)) {
            console.log(`Audio file missing for hotkey ID: ${id}`);
            return;
        }

        const appConfig = await this.settingsManager.getAppConfig();
        console.log(`Playing hotkey [${entry.hotkey}]: "${entry.text}"`);
        await this.audioService.play(
            filePath,
            appConfig.playback,
            appConfig.playbackDevice,
            appConfig.volume,
            appConfig.monitoring,
            appConfig.monitoringDevice,
            appConfig.monitoringVolume
        );
    }
    
    /**
     * Deletes a hotkey, shifts IDs of subsequent hotkeys to fill gaps,
     * renames files, and persists metadata.
     * @param id Identifier of the hotkey to delete.
     */
    public async deleteHotkey(id: number): Promise<void> {
        const index = this.entries.findIndex(e => e.id === id);
        if (index === -1) {
            console.log(`Hotkey ID [${id}] not found.`);
            return;
        }

        // Remove the entry
        this.entries.splice(index, 1);

        // Delete the associated file
        const fileToDelete = path.join(this.hotkeyedDir, `hotkey_${id}.wav`);
        if (existsSync(fileToDelete)) {
            try {
                await fs.unlink(fileToDelete);
            } catch (err: any) {
                console.error(`Failed to delete file ${fileToDelete}:`, err.message);
            }
        }

        // Shift remaining IDs down and rename files to maintain contiguous mapping.
        // Prevents internal fragmented gaps.
        for (const entry of this.entries) {
            if (entry.id > id) {
                const oldId = entry.id;
                const newId = oldId - 1;
                entry.id = newId;

                const oldFile = path.join(this.hotkeyedDir, `hotkey_${oldId}.wav`);
                const newFile = path.join(this.hotkeyedDir, `hotkey_${newId}.wav`);
                if (existsSync(oldFile)) {
                    try {
                        await fs.rename(oldFile, newFile);
                    } catch (err: any) {
                        console.error(`Failed to rename ${oldFile} to ${newFile}:`, err.message);
                    }
                }
            }
        }

        await this.save();
        console.log(`Deleted hotkey ID [${id}] and shifted remaining IDs.`);
    }

    /**
     * Clears all hotkey entries and deletes their respective cached `.wav` files.
     */
    public async clearHotkeys(): Promise<void> {
        this.entries = [];
        await this.save();
        
        try {
            if (existsSync(this.hotkeyedDir)) {
                const files = await fs.readdir(this.hotkeyedDir);
                for (const file of files) {
                    if (file.startsWith('hotkey_') && file.endsWith('.wav')) {
                        await fs.unlink(path.join(this.hotkeyedDir, file));
                    }
                }
            }
            console.log('All hotkeys cleared.');
        } catch (error: any) {
            console.error('Failed to clear hotkey files:', error.message);
        }
    }

    /**
     * Gets the list of loaded hotkey entries.
     */
    public listHotkeys(): HotkeyEntry[] {
        return this.entries;
    }

    /**
     * Prints active hotkey combinations to stdout.
     */
    public printHotkeys(): void {
        if (this.entries.length === 0) {
            console.log('No hotkeys assigned yet.');
        } else {
            console.log('Assigned Hotkeys:');
            this.entries.forEach(h => {
                console.log(`  [${h.id}] ${h.hotkey} -> "${h.text}"`);
            });
        }
    }
}

File: src/core/hotkey_manager.py
"""Hotkey manager for Moon-TTS application.

Allows users to bind keyboard shortcuts to pre-generated TTS audio
clips for instant playback.
"""

from __future__ import annotations

import json
import os
import threading
from pathlib import Path
from typing import Any, TYPE_CHECKING

if TYPE_CHECKING:
    from src.core.settings_manager import SettingsManager


# Type alias for hotkey entry dicts
HotkeyEntry = dict[str, Any]  # keys: id (str), text (str), hotkey (str)


class HotkeyManager:
    """Manages hotkey-bound TTS audio entries.

    Each entry maps a keyboard shortcut to a pre-generated WAV file so
    the user can trigger instant playback without re-synthesising.
    Entries are stored as ``hotkey_{id}.wav`` alongside a
    ``hotkeys.json`` metadata file.
    """

    MAX_ENTRIES: int = 20

    def __init__(
        self,
        tts_service: Any,
        audio_service: Any,
        settings_manager: "SettingsManager",
    ) -> None:
        """Initialize the hotkey manager.

        Args:
            tts_service: Service capable of generating TTS audio files.
            audio_service: Service used to play back audio files.
            settings_manager: Application settings manager instance.
        """
        self._tts_service = tts_service
        self._audio_service = audio_service
        self._settings_manager = settings_manager

        app_dir = settings_manager.get_app_directory()
        self._hotkeyed_dir: Path = Path(app_dir) / "audio" / "hotkeyed"
        self._json_path: Path = self._hotkeyed_dir / "hotkeys.json"

        self._lock = threading.RLock()
        self._entries: list[HotkeyEntry] = []
        self._on_hotkeys_changed_callback = None

    def set_hotkeys_changed_callback(self, callback) -> None:
        """Set a callback to be invoked when hotkeys change."""
        self._on_hotkeys_changed_callback = callback

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def init(self) -> None:
        """Create the hotkeyed directory if needed and load entries."""
        import logging
        logger = logging.getLogger(__name__)
        
        os.makedirs(self._hotkeyed_dir, exist_ok=True)
        self._entries = self._load_entries()
        
        # If entries is empty but there are files, clear legacy/orphaned WAVs
        if not self._entries:
            for file in os.listdir(self._hotkeyed_dir):
                if file.endswith(".wav"):
                    try:
                        os.remove(self._hotkeyed_dir / file)
                    except Exception as e:
                        logger.debug(f"Failed to remove orphaned hotkey wav {file}: {e}")

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def assign_hotkey(self, hotkey: str, text: str) -> None:
        """Assign or update a hotkey binding.

        If an entry with the same *hotkey* already exists its text and
        audio are updated in-place. Otherwise a new entry is created
        (up to ``MAX_ENTRIES``).

        Args:
            hotkey: Keyboard shortcut string (e.g. ``"Ctrl+Shift+1"``).
            text: Text to synthesise and bind to *hotkey*.
        """
        with self._lock:
            # Check for existing entry with same hotkey
            existing = next((e for e in self._entries if e["hotkey"] == hotkey), None)
    
            if existing is not None:
                entry_id = existing["id"]
                existing["text"] = text
            else:
                if len(self._entries) >= self.MAX_ENTRIES:
                    return  # cap reached
                import uuid
                entry_id = uuid.uuid4().hex[:8]
                self._entries.append({"id": entry_id, "text": text, "hotkey": hotkey})
    
            # Generate audio file
            wav_path = str(self._wav_path(entry_id))
            self._tts_service.generate_to_file(text, wav_path)
    
            self._save_entries()
            
            if self._on_hotkeys_changed_callback:
                self._on_hotkeys_changed_callback()

    def play_hotkey(self, entry_id: str) -> None:
        """Play the audio bound to a hotkey entry.

        Args:
            entry_id: String UUID of the hotkey entry.
        """
        wav = self._wav_path(entry_id)
        if not os.path.exists(wav):
            return

        config = self._settings_manager.get_app_config()
        self._audio_service.play_with_config(str(wav), config)

    def delete_hotkey(self, entry_id: str) -> None:
        """Delete a hotkey entry and its corresponding file.

        Args:
            entry_id: String UUID of the entry to remove.
        """
        with self._lock:
            # Find index of entry
            idx = next((i for i, e in enumerate(self._entries) if e["id"] == entry_id), -1)
            if idx == -1:
                return
    
            # Remove the WAV file for the deleted entry
            target = self._wav_path(entry_id)
            if os.path.exists(target):
                os.remove(target)
    
            # Remove entry from list
            self._entries.pop(idx)
    
            self._save_entries()
            
            if self._on_hotkeys_changed_callback:
                self._on_hotkeys_changed_callback()

    def clear_hotkeys(self) -> None:
        """Remove all hotkey entries and their audio files."""
        with self._lock:
            for entry in self._entries:
                wav = self._wav_path(entry["id"])
                if os.path.exists(wav):
                    os.remove(wav)
    
            self._entries.clear()
            self._save_entries()
            
            if self._on_hotkeys_changed_callback:
                self._on_hotkeys_changed_callback()

    def list_hotkeys(self) -> list[HotkeyEntry]:
        """Return a copy of all current hotkey entries.

        Returns:
            List of dicts with keys ``id``, ``text``, ``hotkey``.
        """
        with self._lock:
            return list(self._entries)

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _wav_path(self, entry_id: str) -> Path:
        """Return the WAV file path for an entry.

        Args:
            entry_id: String UUID of the entry.

        Returns:
            Full path to ``phrase_{entry_id}.wav``.
        """
        return self._hotkeyed_dir / f"phrase_{entry_id}.wav"

    def _load_entries(self) -> list[HotkeyEntry]:
        """Read hotkeys.json and return the entry list.

        Returns:
            List of hotkey entry dicts, or empty list on error.
        """
        with self._lock:
            if not self._json_path.exists():
                return []
            try:
                with open(self._json_path, "r", encoding="utf-8") as fh:
                    data = json.load(fh)
                    if not isinstance(data, list):
                        return []
                    
                    # Drop legacy integer-based IDs
                    if any(isinstance(e.get("id"), int) for e in data):
                        return []
                        
                    return data
            except (json.JSONDecodeError, ValueError):
                return []

    def _save_entries(self) -> None:
        """Persist the current entries list to hotkeys.json."""
        with self._lock:
            os.makedirs(self._hotkeyed_dir, exist_ok=True)
            with open(self._json_path, "w", encoding="utf-8") as fh:
                json.dump(self._entries, fh, indent=2, ensure_ascii=False)
